agreement/player.go

```diff
@@ -266,6 +276,74 @@ func (p *player) handleCheckpointEvent(r routerHandle, e checkpointEvent) []acti
 		}}
 }
 
+// updateCredentialArrivalHistory is called at the end of a successful
+// uninterrupted round (just after ensureAction is generated) to collect
+// credential arrival times to dynamically set the filter timeout.
+// It returns the time of the lowest credential's arrival, if one was
+// collected and added to lowestCredentialArrivals, or zero otherwise.
+func (p *player) updateCredentialArrivalHistory(r routerHandle, ver protocol.ConsensusVersion) time.Duration {
+	// only append to lowestCredentialArrivals if this was a successful round completing in period 0.
+	if p.Period != 0 {
+		return 0
+	}
+	// look up the validatedAt time of the winning proposal-vote
+	re := readLowestEvent{T: readLowestVote, Round: p.Round, Period: p.Period}
+	re = r.dispatch(*p, re, proposalMachineRound, p.Round, p.Period, 0).(readLowestEvent)
+	if !re.Filled {
+		return 0
+	}
+
+	p.lowestCredentialArrivals = append(p.lowestCredentialArrivals, re.Vote.validatedAt)
+
+	// trim history to the last dynamicFilterCredentialArrivalHistory samples.
+	if len(p.lowestCredentialArrivals) > dynamicFilterCredentialArrivalHistory {
+		p.lowestCredentialArrivals = p.lowestCredentialArrivals[len(p.lowestCredentialArrivals)-dynamicFilterCredentialArrivalHistory:]
+	}
```

Comment:
could be a fixed size circular buffer to prevent reallocs

Reply:
changed